metadata/src/main/java/org/apache/kafka/image/ClientQuotasImage.java

```diff
@@ -48,11 +48,36 @@
  * <p>
  * This class is thread-safe.
  */
-public record ClientQuotasImage(Map<ClientQuotaEntity, ClientQuotaImage> entities) {
+public final class ClientQuotasImage {
     public static final ClientQuotasImage EMPTY = new ClientQuotasImage(Map.of());
 
+    private final Map<ClientQuotaEntity, ClientQuotaImage> entities;
+
+    // Map from entity type to entity name to set of entries. The entity type could be "user", "client-id", and "ip".
+    // {
+    //   "user": { "user1": [entry1, entry2], "user2": [entry3] },
+    //   "client-id": { "client-id1": [entry4], "client-id2": [entry5] },
+    //   "ip": { "ip1": [entry6], "ip2": [entry7] }
+    // }
+    private final Map<String, Map<String, Set<Entry<ClientQuotaEntity, ClientQuotaImage>>>> entitiesByType;
```

Comment:
just curious, it there a specific reason not to use Map<> ?

```suggestion
    private final Map<String, Map<String, Map<ClientQuotaEntity, ClientQuotaImage>>> entitiesByType;
```

Reply:
Thanks for the suggestion. Updated it.